middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+const nearAPI = require('near-api-js');
+const crypto = require('crypto');
+const nacl = require('tweetnacl');
+const { creatorKeyJson } = require('./near');
+const { sendSms } = require('../utils/sms');
+const { sendMail } = require('../utils/email');
+const models = require('../models');
+const Sequelize = require('sequelize');
+const Op = Sequelize.Op;
+const password = require('secure-random-password');
+const SECURITY_CODE_DIGITS = 6;
+const twoFactorMethods = ['2fa-email', '2fa-phone'];
+const viewMethods = [];
+const changeMethods = ['confirm'];
+
+const getContract = async (contractName, secretKey) => {
+    const keyStore = {
+        async getKey() {
+            return nearAPI.KeyPair.fromString(secretKey);
+        },
+    };
+    const near = await nearAPI.connect({
+        deps: { keyStore },
+        masterAccount: creatorKeyJson && creatorKeyJson.account_id,
+        nodeUrl: process.env.NODE_URL
+    });
+    const contractAccount = new nearAPI.Account(near.connection, contractName);
+    const contract = new nearAPI.Contract(contractAccount, contractName, {
+        viewMethods,
+        changeMethods,
+    });
+    return contract;
+};
+const confirmRequest = async (accountId, request_id) => {
+    const key = await getDetermKey(accountId);
+    const contract = await getContract(accountId, key.secretKey);
+    // always parseInt on requestId. requestId is string in db because it could come from url params etc...
+    const res = await contract.confirm({ request_id: parseInt(request_id) }).catch((e) => {
+        return { success: false, error: e };
+    });
+    return { success: !!res, res };
+};
+/********************************
+2FA Key
+********************************/
+// helper: generates a deterministic key based on the accountId
+const getDetermKey = async (accountId) => {
```

Comment:
why not do this in `getKey` of `keyStore`?

Reply:
is there a reason for using keyStore for this?

Reply:
just seems like a logical place for this, `keyStore` is a place where you obtain a key. You won't need separate `keyStore` for every account in that case as well.

Reply:
sounds good, since each accountId requesting could be different I created a method called getKeyStore that returns a keyStore object